File: rust/chunkshop/src/codeparse/langs/python.rs
//! Python source-code symbol extractor via tree-sitter.
//!
//! Mirrors `python/src/chunkshop/codeparse/langs/python.py` (the
//! `extract_symbols` function and the tree-sitter Query pattern at line
//! 170). Output `Symbol`s feed `SymbolAwareChunker` and have identical
//! `fqn` / `node_id` to Python's emission for the same input.

use crate::codeparse::{build_fqn, Symbol};

/// Tree-sitter tags-style query for Python. Captures function defs, class
/// defs, and method defs (functions nested inside classes).
const PYTHON_TAGS_QUERY: &str = r#"
(function_definition
  name: (identifier) @function.name) @function.def

(class_definition
  name: (identifier) @class.name) @class.def

(class_definition
  body: (block
    (function_definition
      name: (identifier) @method.name) @method.def))
"#;

/// Extract symbols from Python source. Returns symbols with `fqn` +
/// `parent_name` set; `node_id` derivation is up to the caller (the
/// chunker stamps it onto chunk metadata).
///
/// Note: tree-sitter is error-tolerant — it returns a (partial) tree for
/// malformed Python. The chunker layer is responsible for falling back to
/// `sentence_aware` via `has_syntax_errors` (see Task 14).
pub fn extract_symbols(file_path: &str, source: &str) -> Vec<Symbol> {
    use tree_sitter::{Parser, Query, QueryCursor, StreamingIterator};

    let mut parser = Parser::new();
    let language = tree_sitter_python::LANGUAGE.into();
    if parser.set_language(&language).is_err() {
        return Vec::new();
    }
    let Some(tree) = parser.parse(source.as_bytes(), None) else {
        return Vec::new();
    };
    let root = tree.root_node();

    let Ok(query) = Query::new(&language, PYTHON_TAGS_QUERY) else {
        return Vec::new();
    };
    let mut cursor = QueryCursor::new();

    let mut symbols: Vec<Symbol> = Vec::new();
    let source_bytes = source.as_bytes();

    let mut matches = cursor.matches(&query, root, source_bytes);
    while let Some(m) = matches.next() {
        for capture in m.captures {
            let capture_name = query.capture_names()[capture.index as usize];
            let node = capture.node;
            let name = node
                .utf8_text(source_bytes)
                .unwrap_or("")
                .to_string();

            // Python's stdlib-ast walker explicitly does NOT descend into
            // function bodies (it returns after each FunctionDef visit), so
            // any def/class nested inside a function body is invisible to
            // Python's chunker. Mirror that: skip captures whose nearest
            // enclosing definition is a function (def / async def / lambda).
            // Covers all three nesting flavors in real code:
            //   - inner def (closure / helper)         e.g. `_consolidator._bundled`
            //   - inner class (factory / configurator) e.g. `cli._JsonFormatter`
            //   - method on an inner class             (symmetry; same root cause)
            if is_inside_function(node) {
                continue;
            }

            let (symbol_type, parent_name) = match capture_name {
                "function.name" => {
                    // Skip if this is also a method (already captured below).
                    if is_inside_class(node) {
                        continue;
                    }
                    ("function", None)
                }
                "class.name" => ("class", None),
                "method.name" => {
                    let parent = enclosing_class_name(node, source_bytes);
                    ("method", parent)
                }
                _ => continue,
            };

            let fqn = build_fqn(file_path, &name, parent_name.as_deref());
            let line_start = node.start_position().row as u32 + 1;
            let line_end = node.end_position().row as u32 + 1;

            symbols.push(Symbol {
                name,
                fqn,
                symbol_type: symbol_type.to_string(),
                line_start,
                line_end,
                parent_name,
            });
        }
    }

    symbols
}

/// Returns `true` if tree-sitter's parse tree contains any ERROR or MISSING
/// nodes. Mirror of Python's `ast.parse(content)` SyntaxError check at
/// `python/src/chunkshop/chunkers/symbol_aware.py:120-132`. Used by the
/// chunker (Task 14) to trigger fallback to sentence_aware.
pub fn has_syntax_errors(source: &str) -> bool {
    use tree_sitter::Parser;

    let mut parser = Parser::new();
    let language = tree_sitter_python::LANGUAGE.into();
    if parser.set_language(&language).is_err() {
        return true;  // Can't parse → treat as error
    }
    parser
        .parse(source.as_bytes(), None)
        .map(|tree| tree.root_node().has_error())
        .unwrap_or(true)
}

fn is_inside_class(node: tree_sitter::Node) -> bool {
    let mut current = node.parent();
    while let Some(p) = current {
        if p.kind() == "class_definition" {
            return true;
        }
        current = p.parent();
    }
    false
}

/// Returns `true` if `node` is nested inside a function body.
///
/// The capture is an `identifier` whose immediate parent is the
/// `function_definition` / `class_definition` it names — we skip past that
/// owning definition and check from its grandparent up. A symbol nested in
/// a function (closure / inner class / inner def) is invisible to Python's
/// stdlib-ast walker because it returns early after each function visit
/// (see `python/src/chunkshop/codeparse/langs/python.py:122`).
fn is_inside_function(node: tree_sitter::Node) -> bool {
    // Skip the immediate parent (the def/class this identifier IS the name
    // of) — we want to know whether the *owning* definition is itself nested
    // inside another function body, not whether the identifier sits inside
    // the def it names.
    let Some(owner) = node.parent() else { return false; };
    let mut current = owner.parent();
    while let Some(p) = current {
        let k = p.kind();
        if k == "function_definition" || k == "lambda" {
            return true;
        }
        current = p.parent();
    }
    false
}

fn enclosing_class_name(node: tree_sitter::Node, source: &[u8]) -> Option<String> {
    let mut current = node.parent();
    while let Some(p) = current {
        if p.kind() == "class_definition" {
            let name_node = p.child_by_field_name("name")?;
            return name_node.utf8_text(source).ok().map(String::from);
        }
        current = p.parent();
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_top_level_function() {
        let src = "def hello():\n    pass\n";
        let syms = extract_symbols("test.py", src);
        assert_eq!(syms.len(), 1);
        assert_eq!(syms[0].name, "hello");
        assert_eq!(syms[0].symbol_type, "function");
        assert_eq!(syms[0].parent_name, None);
        assert_eq!(syms[0].fqn, "test.hello");
    }

    #[test]
    fn extracts_class_and_methods() {
        let src = "class Foo:\n    def bar(self):\n        pass\n    def baz(self):\n        pass\n";
        let syms = extract_symbols("test.py", src);
        let names: Vec<&str> = syms.iter().map(|s| s.name.as_str()).collect();
        assert!(names.contains(&"Foo"), "expected Foo in {names:?}");
        assert!(names.contains(&"bar"), "expected bar in {names:?}");
        assert!(names.contains(&"baz"), "expected baz in {names:?}");

        let method_bar = syms.iter().find(|s| s.name == "bar").unwrap();
        assert_eq!(method_bar.symbol_type, "method");
        assert_eq!(method_bar.parent_name.as_deref(), Some("Foo"));
        assert_eq!(method_bar.fqn, "test.Foo.bar");
    }

    #[test]
    fn empty_source_returns_no_symbols() {
        let syms = extract_symbols("empty.py", "");
        assert!(syms.is_empty());
    }

    // --- Nested-in-function parity regression suite ---
    //
    // Real-corpus parity check (RM-C follow-up against chunkshop's own
    // python/src/chunkshop tree) surfaced that the Rust extractor emitted
    // 11 extra symbols Python doesn't see. Root cause: Python's stdlib-ast
    // walker doesn't descend into function bodies. These tests lock the fix.

    #[test]
    fn skips_nested_function_inside_function() {
        // Pattern: closure / inner helper. Real example:
        // chunkshop.chunkers._consolidator.build_consolidator defines `_bundled`.
        let src = "\
def outer():
    def inner():
        return 1
    return inner
";
        let syms = extract_symbols("test.py", src);
        let names: Vec<&str> = syms.iter().map(|s| s.name.as_str()).collect();
        assert_eq!(names, vec!["outer"], "expected only outer, got {names:?}");
    }

    #[test]
    fn skips_nested_class_inside_function() {
        // Pattern: class defined inside a function. Real example:
        // chunkshop.cli setup_logging defines `_JsonFormatter`.
        let src = "\
def setup_logging():
    class _JsonFormatter:
        def format(self, record):
            return ''
    return _JsonFormatter
";
        let syms = extract_symbols("test.py", src);
        let names: Vec<&str> = syms.iter().map(|s| s.name.as_str()).collect();
        assert_eq!(names, vec!["setup_logging"], "expected only setup_logging, got {names:?}");
    }

    #[test]
    fn keeps_module_level_class_with_nested_method_using_inner_class() {
        // Outer class + method that uses an inner class inside its body.
        // Outer class + method are at module / class scope → kept.
        // Inner class + its method are inside a function body → skipped.
        let src = "\
class Outer:
    def make(self):
        class Inner:
            def go(self):
                return 1
        return Inner
";
        let syms = extract_symbols("test.py", src);
        let names: Vec<&str> = syms.iter().map(|s| s.name.as_str()).collect();
        // Order may vary; assert set membership.
        let want: std::collections::HashSet<&str> = ["Outer", "make"].into_iter().collect();
        let got: std::collections::HashSet<&str> = names.iter().copied().collect();
        assert_eq!(got, want, "expected {want:?}, got {got:?}");
    }

    #[test]
    fn has_syntax_errors_detects_unbalanced() {
        // Missing close paren — tree-sitter will produce ERROR nodes
        assert!(has_syntax_errors("def hello(\n    pass\n"));
    }

    #[test]
    fn has_syntax_errors_negative_on_valid() {
        assert!(!has_syntax_errors("def hello():\n    pass\n"));
    }
}

File: python/src/chunkshop/codeparse/langs/python.py
"""Python symbol + call-site extraction via tree-sitter.

Lifts the per-language extractor pattern from yonk-full-stack-mig-srv
``edge_queries/python.py`` with two refinements:

1. ``tree_sitter_python`` is imported lazily on first parse so plain
   ``import chunkshop.codeparse`` stays light.
2. Returns chunkshop's pydantic ``ParseResult`` (with ``call_sites``
   embedded) instead of the migration server's three-tuple. Downstream
   chunkers / extractors can consume one object.

On any import or query failure the caller (``tree_sitter_wrapper``) drops
through to the regex fallback. That's the contract: this module is best-
effort and signals failure by raising — never by returning a bogus
ParseResult.
"""
from __future__ import annotations

from typing import Any, Optional

from chunkshop.codeparse.base import CallSite, ParseResult, Symbol
from chunkshop.codeparse.fqn import build_fqn
from chunkshop.codeparse.id import code_symbol_node_id

_PARSER_NAME = "tree-sitter-python"

_CALL_QUERY = (
    "(call function: (identifier) @callee_name) @call\n"
    "(call function: (attribute attribute: (identifier) @callee_name)) @call"
)


def parse(
    *,
    source: bytes,
    file_path: str,
    project_id: str = "default",
) -> ParseResult:
    """Parse Python source via tree-sitter. Raises on import or parse error.

    The "raises on failure" contract is what lets the wrapper layer cleanly
    fall through to regex without inspecting return shapes.
    """
    import tree_sitter_python  # local import: lazy per chunkshop convention

    from tree_sitter import Language, Parser  # type: ignore[import-untyped]

    language = Language(tree_sitter_python.language())
    parser = Parser(language)
    tree = parser.parse(source)
    root = tree.root_node

    symbols = _walk_symbols(root, file_path, source)
    call_sites = _extract_call_sites(
        root=root,
        language=language,
        symbols=symbols,
        file_path=file_path,
        project_id=project_id,
        source=source,
    )
    imports = _extract_imports(root, source)
    return ParseResult(
        symbols=symbols,
        call_sites=call_sites,
        imports=imports,
        language="python",
        parser="tree-sitter",
    )


def _walk_symbols(root: Any, file_path: str, source: bytes) -> list[Symbol]:
    """Collect classes / functions / methods from the tree.

    A ``function_definition`` nested inside a ``class_definition`` is a
    ``method`` with ``parent_name`` = the enclosing class. Top-level
    ``function_definition`` is a ``function``. Nested functions inside
    other functions are NOT emitted — SP-A scope is one level deep, same
    as the migration server.
    """
    symbols: list[Symbol] = []

    def _span(node: Any) -> tuple[int, int]:
        """1-based inclusive span, widened to the decorator block if any.

        A decorated ``def``/``class`` is wrapped in a ``decorated_definition``
        node whose first child is the ``@decorator``. Without this, the span
        would start at the ``def``/``class`` line and silently drop the
        decorator lines from the symbol's ``original_content``.
        """
        span_node = node
        parent = node.parent
        if parent is not None and parent.type == "decorated_definition":
            span_node = parent
        return (span_node.start_point[0] + 1, span_node.end_point[0] + 1)

    def visit(node: Any, parent_class: Optional[str]) -> None:
        ntype = node.type
        if ntype == "class_definition":
            name_node = node.child_by_field_name("name")
            if name_node is not None:
                name = source[name_node.start_byte : name_node.end_byte].decode(
                    errors="replace"
                )
                symbols.append(
                    Symbol(
                        name=name,
                        fqn=build_fqn(file_path, name, None),
                        symbol_type="class",
                        line_start=_span(node)[0],
                        line_end=_span(node)[1],
                        parent_name=None,
                    )
                )
                for child in node.children:
                    visit(child, name)
                return
        elif ntype == "function_definition":
            name_node = node.child_by_field_name("name")
            if name_node is not None:
                name = source[name_node.start_byte : name_node.end_byte].decode(
                    errors="replace"
                )
                sym_type = "method" if parent_class is not None else "function"
                symbols.append(
                    Symbol(
                        name=name,
                        fqn=build_fqn(file_path, name, parent_class),
                        symbol_type=sym_type,
                        line_start=_span(node)[0],
                        line_end=_span(node)[1],
                        parent_name=parent_class,
                    )
                )
                # Don't descend — SP-A is one level deep.
                return

        for child in node.children:
            visit(child, parent_class)

    visit(root, None)
    return symbols


def _enclosing_function(node: Any, source: bytes) -> Optional[tuple[str, Optional[str]]]:
    """Return the OUTERMOST enclosing function/method as (name, parent_class).

    Mirrors :func:`_walk_symbols`' "one level deep" emission rule: nested
    functions are never emitted as Symbols, so a call inside one must be
    attributed to the outermost enclosing function (the symbol that WAS
    emitted), not the innermost. Returns None for a module-level call (no
    enclosing function — there is no symbol to attribute to).
    """
    cur = node.parent
    outermost: Optional[Any] = None
    while cur is not None:
        if cur.type == "function_definition":
            outermost = cur  # keep climbing; the last one wins (highest)
        cur = cur.parent
    if outermost is None:
        return None
    name_node = outermost.child_by_field_name("name")
    if name_node is None:
        return None
    func_name = source[name_node.start_byte : name_node.end_byte].decode(
        errors="replace"
    )
    # parent_class = nearest class_definition ancestor of the outermost
    # function. Because the outermost function has no function ancestor, any
    # enclosing class makes it a method (matching _walk_symbols).
    parent_class: Optional[str] = None
    anc = outermost.parent
    while anc is not None:
        if anc.type == "class_definition":
            cname = anc.child_by_field_name("name")
            if cname is not None:
                parent_class = source[
                    cname.start_byte : cname.end_byte
                ].decode(errors="replace")
            break
        anc = anc.parent
    return (func_name, parent_class)


def _extract_call_sites(
    *,
    root: Any,
    language: Any,
    symbols: list[Symbol],
    file_path: str,
    project_id: str,
    source: bytes,
) -> list[CallSite]:
    try:
        from tree_sitter import Query, QueryCursor  # type: ignore[import-untyped]
    except ImportError:  # pragma: no cover — pinned in [code] extra
        return []
    try:
        q = Query(language, _CALL_QUERY)
    except Exception:  # pragma: no cover — query string is constant
        return []

    intra_file_names = {s.name for s in symbols}
    call_sites: list[CallSite] = []
    cursor = QueryCursor(q)
    captures = cursor.captures(root)
    callee_nodes = captures.get("callee_name", [])

    for callee_node in callee_nodes:
        callee_name = source[
            callee_node.start_byte : callee_node.end_byte
        ].decode(errors="replace")
        enclosing = _enclosing_function(callee_node, source)
        if enclosing is None:
            continue
        func_name, parent_class = enclosing
        caller_fqn = build_fqn(file_path, func_name, parent_class)
        caller_id = code_symbol_node_id(project_id, "python", file_path, caller_fqn)
        line = callee_node.start_point[0] + 1
        snippet = (
            source[
                callee_node.start_byte : min(
                    callee_node.end_byte + 60, len(source)
                )
            ]
            .decode(errors="replace")
            .splitlines()[0]
        )
        call_sites.append(
            CallSite(
                caller_node_id=caller_id,
                callee_name=callee_name,
                line=line,
                snippet=snippet[:240],
                parser=_PARSER_NAME,
                resolved_intra_file=(
                    callee_name in intra_file_names and callee_name != func_name
                ),
            )
        )
    return call_sites


def _extract_imports(root: Any, source: bytes) -> list[str]:
    out: list[str] = []
    for child in root.children:
        if child.type in ("import_statement", "import_from_statement"):
            out.append(
                source[child.start_byte : child.end_byte]
                .decode(errors="replace")
                .strip()
            )
    return out


__all__ = ["parse"]
